pkg/images/zstd.go

```diff
@@ -0,0 +1,272 @@
+	}
+	defer rc.Close()
+
+	return v1.SHA256(rc)
+}
+
+func max(a, b int) int {
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func Zstd(imageName string, dest string, auth authn.Authenticator) (string, error) {
+	var base v1.Image
+	var err error
+
+	fmt.Fprintln(os.Stderr, "fetching metadata for", imageName)
+
+	start := time.Now()
+	base, err = crane.Pull(imageName, crane.WithAuth(auth))
+	if err != nil {
+		return "", fmt.Errorf("pulling %w", err)
+	}
+	fmt.Fprintln(os.Stderr, "pulling took", time.Since(start))
+
+	img, err := zstd(base)
+	if err != nil {
+		return "", err
+	}
+
+	start = time.Now()
+
+	err = crane.Push(img, dest, crane.WithAuth(auth))
+	if err != nil {
+		return "", fmt.Errorf("pushing %s: %w", dest, err)
+	}
+
+	fmt.Fprintln(os.Stderr, "pushing took", time.Since(start))
+
+	d, err := img.Digest()
+	if err != nil {
+		return "", err
+	}
+	image_id := fmt.Sprintf("%s@%s", dest, d)
+	return image_id, nil
+}
+
+func zstd(base v1.Image) (v1.Image, error) {
+
+	// inspired by https://github.com/google/go-containerregistry/blob/v0.15.2/pkg/v1/mutate/mutate.go#L371
+	newImage := empty.Image
+
+	layers, err := base.Layers()
+	if err != nil {
+		return nil, fmt.Errorf("getting layers %w", err)
+	}
+
+	ocf, err := base.ConfigFile()
+	if err != nil {
+		return nil, fmt.Errorf("getting original config file %w", err)
+	}
+
+	addendums := make([]mutate.Addendum, max(len(ocf.History), len(layers)))
+
+	startRecompressing := time.Now()
+
+	// we want nonEmptyHistory to be the v1.History entries that are not EmptyLayer
+	// it should have the same indexes as layers
+	nonEmptyHistory := make([]v1.History, 0, len(layers))
+	for _, h := range ocf.History {
+		if !h.EmptyLayer {
+			nonEmptyHistory = append(nonEmptyHistory, h)
+		}
+	}
+	if len(nonEmptyHistory) != len(layers) {
+		return nil, fmt.Errorf("number of non-empty history entries (%d) is different from number of layers (%d)", len(nonEmptyHistory), len(layers))
+	}
+
+	var historyIdx, addendumIdx int
+	for layerIdx := 0; layerIdx < len(layers); addendumIdx, layerIdx = addendumIdx+1, layerIdx+1 {
+		startLayer := time.Now()
+		compressedSize, err := layers[layerIdx].Size()
+		if err != nil {
+			return nil, fmt.Errorf("getting compressed size: %w", err)
+		}
+		fmt.Fprintln(os.Stderr, "recompressing layer", layerIdx, compressedSize, "created by", nonEmptyHistory[layerIdx].CreatedBy, "with size")
+		newLayer, err := recompressLayer(layers[layerIdx])
+		if err != nil {
+			return nil, fmt.Errorf("setting recompressed layer: %w", err)
+		}
+		// truncate time to 3 decimal places
+		truncTime := time.Duration(int64(time.Since(startLayer).Seconds()*1000)) * time.Millisecond
+		fmt.Fprintln(os.Stderr, "recompressing layer", layerIdx, "took", truncTime)
+		// uncompressedSize, err := newLayer.Size()
+		if err != nil {
+			return nil, fmt.Errorf("getting recompressed size: %w", err)
+		}
+		// compressionRatio := int(float64(compressedSize)/float64(uncompressedSize)*100) / 100.0
```

Comment:
do we want to keep a check for leaving layers with >1 CR  uncompressed, or check for weights layer?